File: scripts/translations/download.js
const fs = require('fs');
const path = require('path');
const requestPromise = require('request-promise');
const twoskyConfig = require('../../.twosky.json')[0];

const { project_id: TWOSKY_PROJECT_ID, languages } = twoskyConfig;
const LOCALES_DIR = '../../client/src/__locales';
const LOCALES_LIST = Object.keys(languages);
const BASE_FILE = 'en.json';
const TWOSKY_URI = process.env.TWOSKY_URI;

/**
 * Prepare params to get translations from twosky
 * @param {string} locale language shortcut
 * @param {object} twosky config twosky
 */
const getRequestUrl = (locale, url, projectId) => {
    return `${url}/download?format=json&language=${locale}&filename=${BASE_FILE}&project=${projectId}`;
};

/**
 * Promise wrapper for writing in file
 * @param {string} filename
 * @param {any} body
 */
function writeInFile(filename, body) {
    let normalizedBody = removeEmpty(JSON.parse(body));

    return new Promise((resolve, reject) => {
        if (typeof normalizedBody !== 'string') {
            try {
                normalizedBody = JSON.stringify(normalizedBody, null, 4); // eslint-disable-line
            } catch (err) {
                reject(err);
            }
        }

        fs.writeFile(filename, normalizedBody, (err) => {
            if (err) reject(err);
            resolve('Ok');
        });
    });
}

/**
 * Clear initial from empty value keys
 * @param {object} initialObject
 */
function removeEmpty(initialObject) {
    let processedObject = {};
    Object.keys(initialObject).forEach(prop => {
        if (initialObject[prop]) {
            processedObject[prop] = initialObject[prop];
        }
    });
    return processedObject;
}

/**
 * Request twosky
 * @param {string} url
 * @param {string} locale
 */
const request = (url, locale) => (
    requestPromise.get(url)
        .then((res) => {
            if (res.length) {
                const pathToFile = path.join(LOCALES_DIR, `${locale}.json`);
                return writeInFile(pathToFile, res);
            }
            return null;
        })
        .then((res) => {
            let result = locale;
            result += res ? ' - OK' : ' - Empty';
            return result;
        })
        .catch((err) => {
            console.log(err);
            return `${locale} - Not OK`;
        }));

/**
 * Sleep.
 * @param {number} ms
 */
const sleep = (ms) => new Promise((resolve) => {
    setTimeout(resolve, ms);
});

/**
 * Download locales
 */
const download = async () => {
    const locales = LOCALES_LIST;

    if (!TWOSKY_URI) {
        console.error('No credentials');
        return;
    }

    const requests = [];
    for (let i = 0; i < locales.length; i++) {
        const locale = locales[i];
        const url = getRequestUrl(locale, TWOSKY_URI, TWOSKY_PROJECT_ID);
        requests.push(request(url, locale));

        // Don't request the Crowdin API too aggressively to prevent spurious
        // 400 errors.
        await sleep(200);
    }

    Promise
        .all(requests)
        .then((res) => {
            res.forEach(item => console.log(item));
        })
        .catch(err => console.log(err));
};

download();

File: scripts/translations/download.go
package main

import (
	"context"
	"flag"
	"fmt"
	"io"
	"log/slog"
	"net/http"
	"net/url"
	"os"
	"path/filepath"
	"slices"
	"sync"
	"time"

	"github.com/AdguardTeam/golibs/errors"
	"github.com/AdguardTeam/golibs/ioutil"
	"github.com/AdguardTeam/golibs/logutil/slogutil"
	"github.com/AdguardTeam/golibs/netutil"
	"github.com/AdguardTeam/golibs/syncutil"
	"github.com/AdguardTeam/golibs/validate"
)

// download and save all translations.
func (c *twoskyClient) download(ctx context.Context, l *slog.Logger) {
	numWorker, err := parseDownloadArgs()
	if err != nil {
		usage(err.Error())
	}

	downloadURI := c.uri.JoinPath("download")

	wg := &sync.WaitGroup{}
	reqCh := make(chan downloadRequest, numWorker)

	dw := &downloadWorker{
		ctx:    ctx,
		l:      l,
		failed: syncutil.NewMap[string, struct{}](),
		client: &http.Client{
			Timeout: 10 * time.Second,
		},
		reqCh: reqCh,
	}

	for range numWorker {
		wg.Go(dw.run)
	}

	for _, baseFile := range c.localizableFiles {
		dir, file := filepath.Split(baseFile)

		for _, lang := range c.langs {
			uri := translationURL(downloadURI, file, c.projectID, lang)

			reqCh <- downloadRequest{
				uri: uri,
				dir: dir,
			}
		}
	}

	close(reqCh)
	wg.Wait()

	printFailedLocales(ctx, l, dw.failed)
}

// parseDownloadArgs parses command-line arguments for the download command.
func parseDownloadArgs() (numWorker int, err error) {
	flagSet := flag.NewFlagSet("download", flag.ExitOnError)
	flagSet.IntVar(&numWorker, "n", 1, "number of concurrent downloads")

	err = flagSet.Parse(os.Args[2:])
	if err != nil {
		// Don't wrap the error since it's informative enough as is.
		return 0, err
	}

	return numWorker, validate.Positive("count", numWorker)
}

// printFailedLocales prints sorted list of failed downloads, if any.  l and
// failed must not be nil.
func printFailedLocales(
	ctx context.Context,
	l *slog.Logger,
	failed *syncutil.Map[string, struct{}],
) {
	var keys []string
	for k := range failed.Range {
		keys = append(keys, k)
	}

	if len(keys) == 0 {
		return
	}

	slices.Sort(keys)

	l.InfoContext(ctx, "failed", "locales", keys)
}

// downloadWorker is a worker for downloading translations.  It uses URLs
// received from the channel to download translations and save them to files.
// Failures are stored in the failed map.  All fields must not be nil.
type downloadWorker struct {
	ctx    context.Context
	l      *slog.Logger
	failed *syncutil.Map[string, struct{}]
	client *http.Client
	reqCh  <-chan downloadRequest
}

// downloadRequest is a request to download a translation.  All fields must not
// be empty.
type downloadRequest struct {
	uri *url.URL
	dir string
}

// run handles the channel of URLs, one by one.  It returns when the channel is
// closed.  It's used to be run in a separate goroutine.
func (w *downloadWorker) run() {
	for req := range w.reqCh {
		q := req.uri.Query()
		code := q.Get("language")

		err := saveToFile(w.ctx, w.l, w.client, req.uri, code, req.dir)
		if err != nil {
			w.l.ErrorContext(w.ctx, "download worker", slogutil.KeyError, err)
			w.failed.Store(code, struct{}{})
		}
	}
}

// saveToFile downloads translation by url and saves it to a file, or returns
// error.
func saveToFile(
	ctx context.Context,
	l *slog.Logger,
	client *http.Client,
	uri *url.URL,
	code string,
	localesDir string,
) (err error) {
	data, err := getTranslation(ctx, l, client, uri.String())
	if err != nil {
		return fmt.Errorf("getting translation %q: %s", code, err)
	}

	if data[len(data)-1] != '\n' {
		data = append(data, '\n')
	}

	name := filepath.Join(localesDir, code+".json")
	err = os.WriteFile(name, data, 0o664)
	if err != nil {
		return fmt.Errorf("writing file: %s", err)
	}

	fmt.Println(name)

	return nil
}

// getTranslation returns received translation data and error.  If err is not
// nil, data may contain a response from server for inspection.  Otherwise, the
// data is guaranteed to be non-empty.
func getTranslation(
	ctx context.Context,
	l *slog.Logger,
	client *http.Client,
	url string,
) (data []byte, err error) {
	resp, err := client.Get(url)
	if err != nil {
		return nil, fmt.Errorf("requesting: %w", err)
	}

	defer slogutil.CloseAndLog(ctx, l, resp.Body, slog.LevelError)

	if resp.StatusCode != http.StatusOK {
		err = fmt.Errorf("url: %q; status code: %s", url, http.StatusText(resp.StatusCode))

		// Go on and download the body for inspection.
	}

	limitReader := ioutil.LimitReader(resp.Body, readLimit.Bytes())

	data, readErr := io.ReadAll(limitReader)
	if readErr != nil {
		return nil, errors.WithDeferred(err, readErr)
	}

	return data, validate.NotEmptySlice("response", data)
}

// translationURL returns a new url.URL with provided query parameters.
func translationURL(baseURL *url.URL, baseFile, projectID string, lang langCode) (uri *url.URL) {
	uri = netutil.CloneURL(baseURL)

	q := uri.Query()
	q.Set("format", "json")
	q.Set("filename", baseFile)
	q.Set("project", projectID)
	q.Set("language", string(lang))

	uri.RawQuery = q.Encode()

	return uri
}
